### ISSUE TITLE
Selecting timestamp type

### ISSUE THREAD

**Issue by supercurio** [diwic/alsa-rs#31]
Hi @diwic!
Thanks a lot for the great work on this crate, it's really useful.

For the [MultiDSP project](https://github.com/supercurio/multidsp/wiki), I'm researching the best way to implement an asynchronous resampler that'll keep multiple sound cards in sync.

The best way seems to be leveraging the PCM status delay, avail and timestamps.
It is already possible to query that today but the timestamps seem to always of the default type: DMA.
For the use case of sound card synchronization, the link time and analog times are useful as well.
Documentation: https://www.kernel.org/doc/html/latest/sound/designs/timestamping.html

I rewrote in Rust [alsa_time.c from alsa-lib](http://git.alsa-project.org/?p=alsa-lib.git;a=blob;f=test/audio_time.c;hb=HEAD) for experimentation and am now testing from that.

Is timestamp type support something you're interested adding?
If I should implement it myself, would you have recommendations?

**Comment by diwic** [comment]
Hi!

Yes, we should add timstamp type support. More specifically, is it the `snd_pcm_status_set_audio_htstamp_config` function you're missing (and the ability to call it before `snd_pcm_status`)?

**Comment by diwic** [comment]
Hmm, so it looks like this is a little less straight forward than usual, because SNDRV_PCM_AUDIO_TSTAMP_TYPE_xxx are not on alsa-lib level (but only on kernel level) and a few functions has not gone into alsa-sys yet...

**Comment by supercurio** [comment]
Nice!
Yes there would be, taking audio_time.c as source:
- snd_pcm_audio_tstamp_config_t
- snd_pcm_audio_tstamp_report_t

These are used both for timestamps and reports.

The later one should be useful for the asynchronous resampler implementation as a challenge is to smooth the right amount the timing data that's a bit noisy.
However, after trying a few kernels, systems and sound cards I haven't seen the report being available yet, but I'll continue looking.

Yes that's right it's not in alsa-sys yet either.
In the kernel doc, it's written:
> ALSA-lib will be modified to make use of STATUS_EXT and effectively deprecate STATUS
It looks indeed like the kernel is ahead of alsa-lib, and that might be a while before the lib is updated on all target systems.

**Comment by supercurio** [comment]
Found it, [this commit in alsa-lib](http://git.alsa-project.org/?p=alsa-lib.git;a=commitdiff;h=b07de7c26b147a19621e35b33fa772de575fa2b8) adds support for timestamp selection.
It's from March 2015, and I also found it in alsa-lib 1.1.6 downloaded from here: ftp://ftp.alsa-project.org/pub/plugins/alsa-plugins-1.1.6.tar.bz2

**Comment by supercurio** [comment]
Issue created on https://github.com/tomaka/cpal/issues/217

**Comment by supercurio** [comment]
Good to know. For this project I'm most interested by activating the delay evaluation to get the analog time.
Ideally that would allow to align the time really precisely between different outputs, I'm not sure yet if even the phase could be matched if the timestamp are precise enough, but that would be really convenient.

**Comment by diwic** [comment]
@supercurio Still around?

Just wanted to give an update that things are moving, but slowly as my time as limited. I've got shared ownership with tomaka on the alsa-sys crate so I can add things without synchronising with others. I've released an alsa-sys with a few new functions and structs in it. Feel free to submit PRs against both alsa-rs or  [my alsa-sys](https://github.com/diwic/alsa-sys) as you fit, or wait for me to get time to do it.
